### PR TITLE
Feature 6208 add nyseliffe exchange to interactive brokers

### PR DESCRIPTION
**Pull request by chriscdev**
#### Description
The following changes were made to add support for NYSELIFFE exchange for Interactive Brokers:
- Added NYSELIFFE to Market.cs and reserved ID 26.
- Added NYSELIFFE to the _futuresExchanges dictionary on InteractiveBrokersBrokerage.cs.
- Added fee support for NYSELIFFE by updating InteractiveBrokersFeeModel.cs, method GetOrderFee().
- Updated symbol-properties-database.csv and added NYSELIFFE futures and options. I wanted to add the indices too but ran into problems, this will be done in a future pull request.
- Updated market-hours-database.json with NYSELIFFE exchange times.
- Added the following margin files to Data\futures\nyseliffe\margins:
  - M1EU.csv
  - M1JP.csv
  - M1MSA.csv
  - MXEA.csv
  - MXEF.csv
  - MXUS.csv
  - YG.csv
  - YI.csv
  - ZG.csv
  - ZI.csv 

Updates after initial review:
- Added method GetGoodFriday() to calculate Good Friday to FuturesExpiryUtilityFunctions.cs.
- Added unit tests for GetGoodFriday().
- Added NYSELIFFE futures expiries to FuturesExpiryFunctions.cs which uses GetGoodFriday() instead of hardcoding the date of Good Friday in the market-hours-database.json.
 - Added unit tests for NYSELIFFE futures expiry.
- Added NYSELIFFE to Exchange.cs
- Removed Equity-nyseliffe-[*] and Index-nyseliffe-[*] from market-hours-database.json
- Updated Option-nyseliffe-[*] to FutureOption-nyseliffe-[*] on market-hours-database.json
- Added the supported NYSELIFFE futureoptions to symbol-properties-database.csv.

#### Related Issue
Closes https://github.com/QuantConnect/Lean/issues/6208

#### Motivation and Context
Require support for NYSE Liffe US (NYSELIFFE) exchange for Interactive Brokers.

#### Requires Documentation Change
Not that I am aware of.

#### How Has This Been Tested?
Updated unit test InteractiveBrokersFuturesTests.CreatesExpectedFuturesContracts to include NYSELIFFE market.
Made sure unit tests passed.
Tested live on Interactive Brokers paper trading account.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Data/symbol-properties/symbol-properties-database.csv [hunk 1]
```diff
@@ -3424,3 +3424,17 @@ ftxus,WBTCUSD,crypto,WBTC/USD,USD,1,1.0,0.0001,WBTC/USD,0.0001
...
+nyseliffe,MXEF,future,MSCI EMERGING MARKETS INDEX,USD,50,0.1,1
+nyseliffe,MXEA,future,MSCI EAFE Index,USD,50,0.1,1
+nyseliffe,M1EU,future,MSCI Europe NTR,USD,10,1,1
+nyseliffe,M1JP,future,The MSCI Japan NTR,USD,10,0.1,1
+nyseliffe,M1MSA,future,MSCI Emerging Markets Asia Net Total Return,USD,100,0.05,1
+nyseliffe,MXUS,future,MSCI USA Index,USD,50,0.1,1
+nyseliffe,YG,future,Mini Sized NY Gold Futures,USD,32.15,0.1,1
+nyseliffe,YI,future,Mini Sized NY Silver Future,USD,1000,0.001,1
+nyseliffe,ZG,future,Gold 100 Troy Oz,USD,100,0.1,1
+nyseliffe,ZI,future,CBOT 5000 Oz Silver Futures,USD,5000,0.001,1
```
Comment (Martin-Molinero) [inline Data/symbol-properties/symbol-properties-database.csv:3438]: Cool! these futures should also be added to the `FuturesExpiryFunctions.cs` & `Futures.cs`

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Welcome to Lean! Thanks for the contribution 🙌 
Leaving a few requests. Can take a look at similar PR https://github.com/QuantConnect/Lean/pull/6190/files adding support for micro futures

File: Common/Market.cs [hunk 2]
```diff
@@ -158,6 +159,11 @@ static Market()
         /// </summary>
         public const string HKFE = "hkfe";
 
+        /// <summary>
+        /// London International Financial Futures and Options Exchange
+        /// </summary>
+        public const string NYSELIFFE = "nyseliffe";
```
[the remark this answers is not in the document]
Comment (chriscdev) [reply Common/Market.cs:165]: Done

File: Data/market-hours/market-hours-database.json [hunk 3]
```diff
@@ -50184,6 +50184,206 @@
...
+          "end": "18:00:00",
+          "state": "market"
+        }
+      ],
+      "saturday": [],
+      "holidays": [],
+      "earlyCloses": {},
+      "lateOpens": {}
+    },
+    "Option-nyseliffe-[*]": {
```
[the remark this answers is not in the document]
Comment (chriscdev) [reply Data/market-hours/market-hours-database.json:50238]: It supports futures and future options. 

I Removed "Equity-nyseliffe-[*]" and "Index-nyseliffe-[*]" from market-hours-database.json and updated "Option-nyseliffe-[*]" to "FutureOption-nyseliffe-[*]", is this correct?

File: Data/future/nyseliffe/margins/M1EU.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+date,initial,maintenance
+20220101,7853,6282
```
[the remark this answers is not in the document]
Comment (chriscdev) [reply Data/future/nyseliffe/margins/M1EU.csv:2]: Got the margins from Interactive Brokers: https://www.interactivebrokers.com/en/index.php?f=1562.
Here is where I got the margin for M1EU: https://contract.ibkr.info/v3.10/index.php?action=Details&site=GEN&conid=522330985
Not sure if I need to add these margin files as I saw that other futures have margin files although most of the file's last entry was in 2020?

File: Data/symbol-properties/symbol-properties-database.csv (hunk 1 above)
Comment (chriscdev) [reply Data/symbol-properties/symbol-properties-database.csv:3438]: Done. 

TL;DR - Good Friday is the only date which will interfere with the expiry date, hence I added code to calculate Good Friday instead of hardcoding the date in market-hours-database.json.

Long version - The NYSELIFFE exchange (which was bought over by ICE) for the added futures is only officially closed for four days of the year, New Year's day, Good Friday, Christmas and Boxing day. The index futures expire on the 3rd Friday of the contract month and metal futures on the 3rd last day of the contract month. Doing some calcs the only date that can affect the expiry date is Good Friday due to the date not being static but changing. Good Friday is deterministic and I have used the code before in other projects to calculate it effectively. This was added to FuturesExpiryUtilityFunctions for others to use as it is way more efficient than adding Good Friday date to the exchange holiday dates array. You will see it in the new commit.

**Review by Martin-Molinero — COMMENTED** [review]
Cool, almost there! Leaving a couple of final requests, will also need to perform a rebase from `master` 👍

File: Data/symbol-properties/symbol-properties-database.csv (hunk 1 above)
Comment (chriscdev) [reply Data/symbol-properties/symbol-properties-database.csv:3438]: Done, added the dates for 2022 and 2023 to the market-hours-database.json

File: Data/market-hours/market-hours-database.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (chriscdev) [reply Data/market-hours/market-hours-database.json:50238]: Removed "FutureOption-nyseliffe-[]" from market-hours-database.json
Added  to FuturesOptionsSymbolMappings.cs: 
            { "YG", "OYG" },
            { "ZG", "OZG" },
            { "ZI", "OZI" }

File: Common/Market.cs [hunk 5]
```diff
@@ -43,7 +43,7 @@ public static class Market
             Tuple.Create(CBOT, 8),
             Tuple.Create(ICE, 9),
             Tuple.Create(CBOE, 10),
-            Tuple.Create(India, 11),
+            Tuple.Create(India, 11),
```
[the remark this answers is not in the document]
Comment (chriscdev) [reply Common/Market.cs]: Removed the trailing spaces

**Comment by chriscdev** [comment]
Just a general question for the future. In the contribution doc it asks to rebase with master, but then there is also a warning: CAUTION Please note that once you have pushed your branch remotely you MUST NOT rebase!

Does this mean that one's you have pushed your branch you shouldn't use rebase at all which means you need to use merge thereafter to get changes into your branch. Or does it mean that make sure when you commit changes that you rebase every time before you push? 

And sorry based on that warning I didn't use rebase but merged.

**Review by Martin-Molinero — APPROVED** [review]
Thanks! 🚀
